odl/space/base_tensors.py

```diff
@@ -0,0 +1,651 @@
+
+            If ``index`` is an integer, ``value`` must be a scalar.
+
+            If ``index`` is a slice or a sequence of slices, ``value``
+            must be broadcastable to the shape of the slice.
+        """
+        raise NotImplementedError('abstract method')
+
+    def __eq__(self, other):
+        """Return ``self == other``.
+
+        Returns
+        -------
+        equals : bool
+            True if all entries of ``self`` and ``other`` are equal,
+            False otherwise.
+        """
+        raise NotImplementedError('abstract method')
+
+    @property
+    def space(self):
+        """The space to which this tensor belongs."""
+        return self.__space
+
+    @property
+    def shape(self):
+        """Number of elements per axis."""
+        return self.space.shape
+
+    @property
+    def dtype(self):
+        """Data type of each entry."""
+        return self.space.dtype
+
+    @property
+    def order(self):
+        """Data storage order, either ``'C'`` or ``'F'``."""
+        return self.space.order
+
+    @property
+    def size(self):
+        """Total number of entries."""
+        return self.space.size
+
+    @property
+    def ndim(self):
+        """Number of dimensions (axes)."""
+        return self.space.ndim
+
+    def __len__(self):
+        """Return ``len(self)``.
+
+        The length is equal to the number of entries along axis 0.
+        """
+        return len(self.space)
+
+    @property
+    def itemsize(self):
+        """Size in bytes of one tensor entry."""
+        return self.dtype.itemsize
+
+    @property
+    def nbytes(self):
+        """Number of bytes in memory occupied by this tensor."""
+        return self.size * self.itemsize
+
+    def __array__(self, dtype=None):
+        """Return a Numpy array from this tensor.
+
+        Parameters
+        ----------
+        dtype :
+            Specifier for the data type of the output array.
+
+        Returns
+        -------
+        array : `numpy.ndarray`
+        """
+        if dtype is None:
+            return self.asarray()
+        else:
+            return self.asarray().astype(dtype, copy=False)
+
+    def __array_wrap__(self, obj):
+        """Return a new tensor wrapping the array ``obj``.
+
+        Parameters
+        ----------
+        obj : `numpy.ndarray`
+            Array to be wrapped.
+
+        Returns
+        -------
+        vector : `BaseGeneralizedTensor`
+            Tensor wrapping ``obj``.
+        """
+        if obj.ndim == 0:
+            return self.space.field.element(obj)
+        else:
+            return self.space.element(obj)
+
+    def __ne__(self, other):
+        """Return ``self != other``."""
+        return not self.__eq__(other)
+
+    def __str__(self):
+        """Return ``str(self)``."""
+        return arraynd_repr(self)
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        return '{!r}.element(\n{}\n)'.format(self.space,
```

Comment:
It looks quite bad when you have more than 1 dimension and start in the same line. An alternative would be to print as before for `ndim == 1` and use Numpy printing indented by 4 if `ndim > 1`. Needs more experimentation.